src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java

```diff
@@ -335,48 +468,76 @@ public static MetadataTransform replaceMetadata(
           "Neither string or pattern is defined");
     }
 
-    Set<String> keys = new HashSet<String>(parseList(config, "key"));
+    List<Key> keys = keyList(config);
     if (keys.isEmpty()) {
       log.warning("No entries listed to replace");
     }
     return new ReplaceTransform(keys, toMatch, replacementPattern, overwrite);
   }
 
   private static class ReplaceTransform implements MetadataTransform {
-    private final List<String> keys;
+    private final List<Key> keys;
     private final Pattern toMatch;
     private final String replacement;
     private final boolean overwrite;
 
-    public ReplaceTransform(Collection<String> keys, Pattern toMatch,
+    public ReplaceTransform(List<Key> keys, Pattern toMatch,
         String replacement, boolean overwrite) {
-      this.keys = Collections.unmodifiableList(new ArrayList<String>(keys));
+      this.keys = keys;
       this.toMatch = toMatch;
       this.replacement = replacement;
       this.overwrite = overwrite;
     }
 
     @Override
     public void transform(Metadata metadata, Map<String, String> params) {
-      for (String key : keys) {
-        Set<String> original = metadata.getAllValues(key);
-        if (original.isEmpty()) {
-          log.log(Level.FINE, "No values for {0}. Skipping", key);
-          continue;
+      for (Key key : keys) {
+        switch (key.keyset) {
+          case METADATA:
+            replaceInMetadata(key.key, metadata);
+            break;
+          case PARAMS:
+            replaceInParams(key.key, params);
+            break;
         }
-        log.log(Level.FINE, "Replacing values that match {0} with {1}: {2}",
-            new Object[] {toMatch, replacement, original});
-        Set<String> values = new HashSet<String>(original);
-        for (String value : original) {
-          String newValue = toMatch.matcher(value).replaceAll(replacement);
-          if (overwrite) {
-            values.remove(value);
-          }
-          values.add(newValue);
+      }
+    }
+
+    private void replaceInMetadata(String key, Metadata metadata) {
+      Set<String> original = metadata.getAllValues(key);
+      if (original.isEmpty()) {
+        log.log(Level.FINE, "No metadata values for {0}. Skipping", key);
+        return;
+      }
+      log.log(Level.FINE,
+         "Replacing {1} with {2} in metadata values of {0}: {3}",
+          new Object[] {key, toMatch, replacement, original});
+      Set<String> values = new HashSet<String>(original);
+      for (String value : original) {
+        String newValue = toMatch.matcher(value).replaceAll(replacement);
+        if (overwrite) {
+          values.remove(value);
         }
-        log.log(Level.FINE, "After replacing: {0}", values);
-        metadata.set(key, values);
+        values.add(newValue);
```

Comment:
I haven't thought deeply about it, but it seems like the set semantics used for values here could get us into trouble with overwrite. For example, suppose the original values are "w" and "ww", and the pattern replacement is "." to "$0$0" (the use of $ seems undocumented, as goes group 0; what am I missing?). Then "w" => "ww", and "w" is erased. Then "ww" => "www" and "ww" is erased. Now only "www" is left, but we started with two values.